### PR TITLE
add sameness group to source intention

### PR DESCRIPTION
**Pull request by malizz**
Changes proposed in this PR:
- added samenessgroup to source intention
- updated source intention validation 
- updated tests on validation 

How I tested it:
Followed instructions on testing a new CRD: tested docker image and added an intention with samenessgroup. 

Checklist:
- [x] Tests added
- NA CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: control-plane/api/v1alpha1/serviceintentions_types_test.go [hunk 1]
```diff
@@ -1377,8 +1455,8 @@ func TestServiceIntentions_Validate(t *testing.T) {
 			namespacesEnabled: true,
 			partitionsEnabled: true,
 			expectedErrMsgs: []string{
-				`spec.sources[0]: Invalid value: v1alpha1.SourceIntention{Name:"web", Namespace:"namespace-b", Peer:"peer-other", Partition:"partition-other", Action:"allow", Permissions:v1alpha1.IntentionPermissions(nil), Description:""}: Both source.peer and source.partition cannot be set.`,
-				`spec.sources[1]: Invalid value: v1alpha1.SourceIntention{Name:"db", Namespace:"namespace-c", Peer:"peer-2", Partition:"partition-2", Action:"deny", Permissions:v1alpha1.IntentionPermissions(nil), Description:""}: Both source.peer and source.partition cannot be set.`,
+				`spec.sources[0]: Invalid value: v1alpha1.SourceIntention{Name:"web", Namespace:"namespace-b", Peer:"peer-other", Partition:"partition-other", SamenessGroup:"", Action:"allow", Permissions:v1alpha1.IntentionPermissions(nil), Description:""}: cannot set peer and partition at the same time.`,
```
Comment (thisisnotashwin) [inline control-plane/api/v1alpha1/serviceintentions_types_test.go:1458]: ❤️

**Review by thisisnotashwin — APPROVED** [review]
This is excellent!!

File: control-plane/api/v1alpha1/serviceintentions_types.go [hunk 2]
```diff
@@ -285,6 +288,46 @@ func (in *ServiceIntentions) Validate(consulMeta common.ConsulMeta) error {
...
 }
 
+func (in *SourceIntention) validate(path *field.Path, partitionsEnabled bool) field.ErrorList {
+	var errs field.ErrorList
+
+	if in.Name == "" {
+		errs = append(errs, field.Required(path.Child("name"), "name is required."))
+	}
+
+	if strings.Contains(in.Partition, WildcardSpecifier) {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/serviceintentions_types.go:298]: Can you add some additional test coverage for the wildcard cases?

File: control-plane/api/v1alpha1/serviceintentions_types.go [hunk 3]
```diff
@@ -285,6 +288,46 @@ func (in *ServiceIntentions) Validate(consulMeta common.ConsulMeta) error {
 	return nil
 }
 
+func (in *SourceIntention) validate(path *field.Path, partitionsEnabled bool) field.ErrorList {
+	var errs field.ErrorList
+
+	if in.Name == "" {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/serviceintentions_types.go:294]: Can you add some test coverage for this case?

File: control-plane/api/v1alpha1/serviceintentions_types.go [hunk 4]
```diff
@@ -285,6 +288,46 @@ func (in *ServiceIntentions) Validate(consulMeta common.ConsulMeta) error {
...
+
+	if in.SamenessGroup != "" && in.Partition != "" {
+		errs = append(errs, field.Invalid(path, *in, "cannot set samenessgroup and partition at the same time."))
+	}
+
+	if in.SamenessGroup != "" && in.Peer != "" {
+		errs = append(errs, field.Invalid(path, *in, "cannot set samenessgroup and peer at the same time."))
+	}
+
+	if len(in.Description) > metaValueMaxLength {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/serviceintentions_types.go:324]: can you add some test coverage for this case?

File: control-plane/api/v1alpha1/shared_types.go [hunk 5]
```diff
@@ -16,6 +16,8 @@ import (
 
 // This file contains structs that are shared between multiple config entries.
 
+const metaValueMaxLength = 512
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/shared_types.go:19]: Is this max length in bytes? maybe we can add the units? Also is it a common type it's only used by the intentions right?

**Review by wilkermichael — COMMENTED** [review]
Awesome job, just a couple things

File: control-plane/api/v1alpha1/shared_types.go (hunk 5 above)
Comment (malizz) [reply control-plane/api/v1alpha1/shared_types.go:19]: It is the max length of the string size. In Consul it is used by more than just intention meta, any config entry meta should not be more than this length so I added it to shared types to later use in validating other config entries. 
I will add test coverage for this check.

Comment (wilkermichael) [reply control-plane/api/v1alpha1/shared_types.go:19]: ah okay, that's great then. Can you add a comment to that effect? Thanks!

**Review by wilkermichael — APPROVED** [review]

**Pull request merged**